src/utils/config-context.ts

```diff
@@ -1,16 +1,48 @@
-import { createContext, createElement, useContext, FC } from 'react'
+import { createContext, createElement, useContext, useState, FC } from 'react'
 
-import mergeConfig from './merge-config'
-import { SWRConfiguration } from '../types'
+import { defaultProvider } from './config'
+import { initCache } from './cache'
+import { mergeConfigs } from './merge-config'
+import { UNDEFINED } from './helper'
+import {
+  SWRConfiguration,
+  FullConfiguration,
+  ConfigOptions,
+  Cache
+} from '../types'
 
-export const SWRConfigContext = createContext<SWRConfiguration>({})
+export const SWRConfigContext = createContext<Partial<FullConfiguration>>({})
 
 const SWRConfig: FC<{
-  value: SWRConfiguration
+  value?: SWRConfiguration &
+    Partial<ConfigOptions> & {
+      provider?: (cache: Readonly<Cache>) => Cache
+    }
 }> = ({ children, value }) => {
   // Extend parent context values and middlewares.
-  value = mergeConfig(useContext(SWRConfigContext) || {}, value)
-  return createElement(SWRConfigContext.Provider, { value }, children)
+  const extendedConfig = mergeConfigs(useContext(SWRConfigContext), value)
+
+  // Should not use the inherited provider.
+  const provider = value && value.provider
+
+  // Use a lazy initialized state to create the cache on first access.
```

Comment:
Might be unexpected, as a user, that the cache isn't initialized immediately

Reply:
It will be initialized immediately when rendering the `SWRConfig` component. This is critical for React 18+ with concurrent rendering.

Reply:
Oh I misunderstood the comment here 👍 

> create the cache on first access